### PR TITLE
feat(zjsonpatch): move io.fabric8:zjsonpatch as a module in Fabric8 kubernetes Client

### LINKED ISSUE #3489: Critical CVE CVE-2017-7525 in zjsonpatch-0.3.0.jar
The kubernetes-client uses the zjsonpatch-0.3.0.jar library which has a critical vulnerability: https://nvd.nist.gov/vuln/detail/CVE-2017-7525
Please upgrade to an newer version of zjsonpatch.

### LINKED ISSUE #5480: JSONPatch functionality
## Description

Currently, we're leveraging Fabric8 outdated [fork](https://github.com/fabric8io/zjsonpatch) of https://github.com/flipkart-incubator/zjsonpatch.

The fork includes several commits ahead of the forked repository to remove dependencies (one of the principles in the Fabric8 Kubernetes Client is to keep dependencies to the minimum).
This comment provides additional details: https://github.com/fabric8io/kubernetes-client/issues/3489#issuecomment-1361440108

We need to find a proper replacement. we're not affected by security issues because our Jackson dependency is managed by the client project. However, eventually these dependencies might not be compatible any more.

Relates to:
- https://github.com/fabric8io/kubernetes-client/issues/3489
- https://github.com/fabric8io/kubernetes-client/pull/4700

### PR DESCRIPTION
**Pull request by rohanKanojia**
## Description

Fix #5480
Fix #3489

Supersedes closes #4700

Currently, we rely on outdated `io.fabric8:zjsonpatch` dependency instead of the upstream `com.flipkart.zjsonpatch:zjsonpatch` dependency. The main issue is that the upstream repository depends on `org.apache.commons:commons-collections4` .

+ Add a new module `zjsonpatch` that would directly include `com.flipkart.zjsonpatch:zjsonpatch` dependency excluding `org.apache.commons:commons-collections4`.
+ Port required Apache Commons Collections4 method `ListUtils.longestCommonSubsequence` and related
      classes inside zjsonpatch module

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [X] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by rohanKanojia** [comment]
After migrating to the upstream `com.flipkart.zjsonpatch` dependency, KubernetesMutationHookHandlingTest started failing. Upon checking the failures I found that this module depends on [Java Operator SDK](https://github.com/operator-framework/java-operator-sdk/blob/a345797d5dcbec07723507637d6f0210d94b803b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/GenericKubernetesResourceMatcher.java#L189) that uses `io.fabric8:zjsonpatch` library to calculate JSON diff in [GenericKubernetesResourceMatcher](https://github.com/operator-framework/java-operator-sdk/blob/a345797d5dcbec07723507637d6f0210d94b803b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/kubernetes/GenericKubernetesResourceMatcher.java#L189).
    
Adding back the `io.fabric8.zjsonpatch.JsonDiff` fixes the failing test.

File: zjsonpatch/src/main/java/io/fabric8/zjsonpatch/JsonDiff.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+package io.fabric8.zjsonpatch;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+public class JsonDiff {
+  private JsonDiff() {
+  }
+
+  public static JsonNode asJson(final JsonNode source, final JsonNode target) {
+    return com.flipkart.zjsonpatch.JsonDiff.asJson(source, target);
```
Comment (rohanKanojia) [inline zjsonpatch/src/main/java/io/fabric8/zjsonpatch/JsonDiff.java:25]: I had to reintroduce this class due to https://github.com/fabric8io/kubernetes-client/pull/6276#issuecomment-2293363660

**Review by shawkins — APPROVED** [review]
LGTM, thanks @rohanKanojia

File: zjsonpatch/src/main/java/io/fabric8/zjsonpatch/internal/collections4/ListUtils.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * Ported from <a href=
+ * "https://github.com/apache/commons-collections/blob/master/src/main/java/org/apache/commons/collections4/ListUtils.java">Apache
+ * Commons Collections</a>
+ */
+public class ListUtils {
+  private ListUtils() {
+  }
+
+  public static <T> List<T> longestCommonSubsequence(final List<T> list1, final List<T> list2) {
```
Comment (manusa) [inline zjsonpatch/src/main/java/io/fabric8/zjsonpatch/internal/collections4/ListUtils.java:31]: This means that all of the other commons-collections bulky classes are no longer needed? Nice

Comment (rohanKanojia) [reply zjsonpatch/src/main/java/io/fabric8/zjsonpatch/internal/collections4/ListUtils.java:31]: Thanks to ChatGPT ;-)

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Pull request merged**